File: python/ivit/train/exporter.py
"""
Model exporter for converting trained models to deployment formats.

Supports ONNX, TorchScript, and backend-specific formats.
"""

from typing import List, Dict, Any, Optional, Tuple
from pathlib import Path
import numpy as np
import logging

logger = logging.getLogger(__name__)


class ModelExporter:
    """
    Export trained models to various deployment formats.

    Supports:
    - ONNX (cross-platform)
    - TorchScript (PyTorch native)
    - OpenVINO IR (Intel optimization)
    - TensorRT Engine (NVIDIA optimization)

    Args:
        model: PyTorch model to export
        device: Device the model is on

    Examples:
        >>> exporter = ModelExporter(model, device)
        >>> exporter.export("model.onnx", format="onnx")
        >>> exporter.export("model.xml", format="openvino", quantize="int8")
    """

    def __init__(self, model: Any, device: Any):
        self.model = model
        self.device = device

    def export(
        self,
        path: str,
        format: str = "onnx",
        optimize_for: Optional[str] = None,
        quantize: Optional[str] = None,
        input_shape: tuple = (1, 3, 224, 224),
        opset_version: int = 17,
        class_names: Optional[List[str]] = None,
        calibration_data: Optional[List[np.ndarray]] = None,
        **kwargs
    ) -> str:
        """
        Export model to specified format.

        Args:
            path: Output path
            format: Export format ("onnx", "torchscript", "openvino", "tensorrt")
            optimize_for: Target hardware optimization
            quantize: Quantization mode ("fp16", "int8")
            input_shape: Model input shape
            opset_version: ONNX opset version
            class_names: Class names for metadata
            calibration_data: Data for INT8 calibration

        Returns:
            Path to exported model
        """
        format = format.lower()

        # Create output directory
        Path(path).parent.mkdir(parents=True, exist_ok=True)

        if format == "onnx":
            return self._export_onnx(
                path, input_shape, opset_version, quantize, class_names
            )
        elif format == "torchscript":
            return self._export_torchscript(path, input_shape)
        elif format == "openvino":
            return self._export_openvino(
                path, input_shape, quantize, calibration_data
            )
        elif format == "tensorrt":
            return self._export_tensorrt(
                path, input_shape, quantize
            )
        else:
            raise ValueError(f"Unsupported format: {format}")

    def _export_onnx(
        self,
        path: str,
        input_shape: tuple,
        opset_version: int,
        quantize: Optional[str],
        class_names: Optional[List[str]],
    ) -> str:
        """Export to ONNX format."""
        import torch

        self.model.eval()

        # Create dummy input
        dummy_input = torch.randn(input_shape).to(self.device)

        # Get input/output names
        input_names = ["input"]
        output_names = ["output"]

        # Dynamic axes for batch size
        dynamic_axes = {
            "input": {0: "batch_size"},
            "output": {0: "batch_size"},
        }

        # Export to ONNX
        logger.info(f"Exporting to ONNX: {path}")
        torch.onnx.export(
            self.model,
            dummy_input,
            path,
            input_names=input_names,
            output_names=output_names,
            dynamic_axes=dynamic_axes,
            opset_version=opset_version,
            do_constant_folding=True,
        )

        # Apply quantization if requested
        if quantize == "fp16":
            path = self._quantize_onnx_fp16(path)
        elif quantize == "int8":
            logger.warning("INT8 quantization requires calibration data")

        # Save class names
        if class_names:
            labels_path = Path(path).with_suffix(".txt")
            with open(labels_path, 'w') as f:
                f.write('\n'.join(class_names))
            logger.info(f"Saved labels to {labels_path}")

        logger.info(f"ONNX export complete: {path}")
        return path

    def _quantize_onnx_fp16(self, path: str) -> str:
        """Convert ONNX model to FP16."""
        try:
            from onnxconverter_common import float16
            import onnx

            model = onnx.load(path)
            model_fp16 = float16.convert_float_to_float16(model)

            fp16_path = str(Path(path).with_suffix('')) + "_fp16.onnx"
            onnx.save(model_fp16, fp16_path)

            logger.info(f"FP16 quantization complete: {fp16_path}")
            return fp16_path

        except ImportError:
            logger.warning("onnxconverter-common not installed, skipping FP16 conversion")
            return path

    def _export_torchscript(
        self,
        path: str,
        input_shape: tuple,
    ) -> str:
        """Export to TorchScript format."""
        import torch

        self.model.eval()

        # Create dummy input
        dummy_input = torch.randn(input_shape).to(self.device)

        # Trace model
        logger.info(f"Exporting to TorchScript: {path}")
        traced = torch.jit.trace(self.model, dummy_input)
        traced.save(path)

        logger.info(f"TorchScript export complete: {path}")
        return path

    def _export_openvino(
        self,
        path: str,
        input_shape: tuple,
        quantize: Optional[str],
        calibration_data: Optional[List[np.ndarray]],
    ) -> str:
        """Export to OpenVINO IR format."""
        import torch

        # First export to ONNX
        onnx_path = str(Path(path).with_suffix('.onnx'))
        self._export_onnx(onnx_path, input_shape, 17, None, None)

        try:
            import openvino as ov
            from openvino import Core
            from openvino.runtime import serialize

            core = Core()
            logger.info(f"Converting to OpenVINO IR: {path}")

            # Read ONNX model
            model = core.read_model(onnx_path)

            # Apply quantization if requested
            if quantize == "int8" and calibration_data is not None:
                model = self._quantize_openvino_int8(model, calibration_data)
            elif quantize == "fp16":
                # Compress weights to FP16 during serialization
                xml_path = str(Path(path).with_suffix('.xml'))
                ov.save_model(model, xml_path, compress_to_fp16=True)
                logger.info(f"OpenVINO IR export complete (FP16): {xml_path}")
                return xml_path

            # Serialize to IR format
            xml_path = str(Path(path).with_suffix('.xml'))
            serialize(model, xml_path)

            logger.info(f"OpenVINO IR export complete: {xml_path}")
            return xml_path

        except ImportError:
            logger.warning("OpenVINO not installed, returning ONNX path")
            return onnx_path

    def _quantize_openvino_int8(self, model, calibration_data: List[np.ndarray]):
        """Apply INT8 quantization using NNCF."""
        try:
            import nncf

            # Create calibration dataset
            def transform_fn(data_item):
                return np.expand_dims(data_item, 0)

            calibration_dataset = nncf.Dataset(calibration_data, transform_fn)

            # Quantize model
            quantized_model = nncf.quantize(
                model,
                calibration_dataset,
                preset=nncf.QuantizationPreset.MIXED,
            )

            logger.info("INT8 quantization complete")
            return quantized_model

        except ImportError:
            logger.warning("NNCF not installed, skipping INT8 quantization")
            return model

    def _export_tensorrt(
        self,
        path: str,
        input_shape: tuple,
        quantize: Optional[str],
    ) -> str:
        """Export to TensorRT engine format."""
        # First export to ONNX
        onnx_path = str(Path(path).with_suffix('.onnx'))
        self._export_onnx(onnx_path, input_shape, 17, None, None)

        try:
            import tensorrt as trt

            logger.info(f"Converting to TensorRT: {path}")

            # Create builder and network
            TRT_LOGGER = trt.Logger(trt.Logger.WARNING)
            builder = trt.Builder(TRT_LOGGER)
            network = builder.create_network(
                1 << int(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)
            )
            parser = trt.OnnxParser(network, TRT_LOGGER)

            # Parse ONNX
            with open(onnx_path, 'rb') as f:
                if not parser.parse(f.read()):
                    for error in range(parser.num_errors):
                        logger.error(parser.get_error(error))
                    raise RuntimeError("Failed to parse ONNX model")

            # Configure builder
            config = builder.create_builder_config()
            config.set_memory_pool_limit(trt.MemoryPoolType.WORKSPACE, 1 << 30)  # 1GB

            if quantize == "fp16":
                config.set_flag(trt.BuilderFlag.FP16)
                logger.info("Enabled FP16 precision")
            elif quantize == "int8":
                config.set_flag(trt.BuilderFlag.INT8)
                logger.warning("INT8 requires calibrator, falling back to FP16")
                config.set_flag(trt.BuilderFlag.FP16)

            # Build engine
            serialized_engine = builder.build_serialized_network(network, config)

            if serialized_engine is None:
                raise RuntimeError("Failed to build TensorRT engine")

            # Save engine
            engine_path = str(Path(path).with_suffix('.engine'))
            with open(engine_path, 'wb') as f:
                f.write(serialized_engine)

            logger.info(f"TensorRT export complete: {engine_path}")
            return engine_path

        except ImportError:
            logger.warning("TensorRT not installed, returning ONNX path")
            return onnx_path


def export_model(
    model: Any,
    path: str,
    format: str = "onnx",
    input_shape: tuple = (1, 3, 224, 224),
    **kwargs
) -> str:
    """
    Convenience function to export a model.

    Args:
        model: Model to export
        path: Output path
        format: Export format
        input_shape: Input shape
        **kwargs: Additional arguments

    Returns:
        Path to exported model

    Examples:
        >>> from ivit.train import export_model
        >>> export_model(model, "model.onnx", input_shape=(1, 3, 640, 640))
    """
    import torch
    device = next(model.parameters()).device
    exporter = ModelExporter(model, device)
    return exporter.export(path, format, input_shape=input_shape, **kwargs)

File: src/train/exporter.cpp
/**
 * @file exporter.cpp
 * @brief Model exporter implementation
 */

#include "ivit/train/exporter.hpp"

#include <iostream>
#include <fstream>
#include <filesystem>
#include <algorithm>

#ifdef IVIT_HAS_TORCH
#include <torch/torch.h>
#include <torch/script.h>
#endif

namespace ivit {
namespace train {

namespace fs = std::filesystem;

// ============================================================================
// Implementation Details
// ============================================================================

struct ModelExporter::Impl {
#ifdef IVIT_HAS_TORCH
    torch::jit::Module jit_model;
    bool has_model = false;
#endif
};

// ============================================================================
// Construction
// ============================================================================

#ifdef IVIT_HAS_TORCH
ModelExporter::ModelExporter(torch::nn::AnyModule model, torch::Device device)
    : device_(device)
    , impl_(std::make_unique<Impl>())
{
    // Note: AnyModule is tricky to trace, we'll handle this case specially
}
#endif

ModelExporter::ModelExporter(const std::string& checkpoint_path)
    : checkpoint_path_(checkpoint_path)
    , impl_(std::make_unique<Impl>())
{
#ifdef IVIT_HAS_TORCH
    if (!fs::exists(checkpoint_path)) {
        throw std::runtime_error("Checkpoint not found: " + checkpoint_path);
    }

    try {
        impl_->jit_model = torch::jit::load(checkpoint_path);
        impl_->has_model = true;
        device_ = torch::kCPU;
    } catch (const std::exception& e) {
        throw std::runtime_error("Failed to load checkpoint: " + std::string(e.what()));
    }
#else
    throw std::runtime_error("LibTorch not available");
#endif
}

ModelExporter::~ModelExporter() = default;

// ============================================================================
// Export Methods
// ============================================================================

std::string ModelExporter::export_model(const std::string& path, const ExportOptions& options) {
    std::string format = options.format;
    std::transform(format.begin(), format.end(), format.begin(), ::tolower);

    if (format == "onnx") {
        return export_onnx(path, options.input_shape, options.opset_version, options.dynamic_batch);
    } else if (format == "torchscript" || format == "pt") {
        return export_torchscript(path, options.input_shape);
    } else if (format == "openvino") {
        std::vector<cv::Mat> calib_data;  // Empty for now
        return export_openvino(path, options.input_shape, options.quantize, calib_data);
    } else if (format == "tensorrt") {
        return export_tensorrt(path, options.input_shape, options.quantize);
    }

    throw std::invalid_argument("Unsupported export format: " + format);
}

std::string ModelExporter::export_onnx(
    const std::string& path,
    const std::vector<int64_t>& input_shape,
    int opset_version,
    bool dynamic_batch
) {
#ifdef IVIT_HAS_TORCH
    // Create parent directory
    fs::path dir = fs::path(path).parent_path();
    if (!dir.empty() && !fs::exists(dir)) {
        fs::create_directories(dir);
    }

    std::cout << "[ModelExporter] Exporting to ONNX: " << path << std::endl;

    // For ONNX export, we first need to save as TorchScript
    // then convert using external tools
    std::string ts_path = path;
    size_t pos = ts_path.find(".onnx");
    if (pos != std::string::npos) {
        ts_path.replace(pos, 5, "_temp.pt");
    } else {
        ts_path += "_temp.pt";
    }

    // Export TorchScript first
    export_torchscript(ts_path, input_shape);

    std::cout << "[ModelExporter] Note: For full ONNX export, use Python:" << std::endl;
    std::cout << "  import torch" << std::endl;
    std::cout << "  model = torch.jit.load('" << ts_path << "')" << std::endl;
    std::cout << "  torch.onnx.export(model, torch.randn(" << input_shape[0];
    for (size_t i = 1; i < input_shape.size(); ++i) {
        std::cout << ", " << input_shape[i];
    }
    std::cout << "), '" << path << "')" << std::endl;

    // Save class labels
    save_labels(path);

    return ts_path;
#else
    throw std::runtime_error("LibTorch not available");
#endif
}

std::string ModelExporter::export_torchscript(
    const std::string& path,
    const std::vector<int64_t>& input_shape
) {
#ifdef IVIT_HAS_TORCH
    // Create parent directory
    fs::path dir = fs::path(path).parent_path();
    if (!dir.empty() && !fs::exists(dir)) {
        fs::create_directories(dir);
    }

    std::cout << "[ModelExporter] Exporting to TorchScript: " << path << std::endl;

    if (impl_->has_model) {
        // We already have a JIT model, just save it
        impl_->jit_model.save(path);
        std::cout << "[ModelExporter] TorchScript export complete: " << path << std::endl;
    } else {
        throw std::runtime_error("No model available to export");
    }

    // Save class labels
    save_labels(path);

    return path;
#else
    throw std::runtime_error("LibTorch not available");
#endif
}

std::string ModelExporter::export_openvino(
    const std::string& path,
    const std::vector<int64_t>& input_shape,
    const std::string& quantize,
    const std::vector<cv::Mat>& calibration_data
) {
#ifdef IVIT_HAS_TORCH
    // First export to ONNX
    std::string onnx_path = fs::path(path).replace_extension(".onnx").string();
    export_onnx(onnx_path, input_shape, 17, true);

    std::cout << "[ModelExporter] Converting to OpenVINO IR: " << path << std::endl;

    std::cout << "[ModelExporter] To convert to OpenVINO, use:" << std::endl;
    std::cout << "  mo --input_model " << onnx_path << " --output_dir "
              << fs::path(path).parent_path().string() << std::endl;

    if (quantize == "fp16") {
        std::cout << "  Add: --compress_to_fp16" << std::endl;
    } else if (quantize == "int8") {
        std::cout << "  For INT8, use OpenVINO NNCF toolkit" << std::endl;
    }

    return onnx_path;
#else
    throw std::runtime_error("LibTorch not available");
#endif
}

std::string ModelExporter::export_tensorrt(
    const std::string& path,
    const std::vector<int64_t>& input_shape,
    const std::string& quantize
) {
#ifdef IVIT_HAS_TORCH
    // First export to ONNX
    std::string onnx_path = fs::path(path).replace_extension(".onnx").string();
    export_onnx(onnx_path, input_shape, 17, true);

    std::cout << "[ModelExporter] Converting to TensorRT: " << path << std::endl;

    std::cout << "[ModelExporter] To convert to TensorRT, use:" << std::endl;
    std::cout << "  trtexec --onnx=" << onnx_path
              << " --saveEngine=" << fs::path(path).replace_extension(".engine").string();

    if (quantize == "fp16") {
        std::cout << " --fp16";
    } else if (quantize == "int8") {
        std::cout << " --int8";
    }
    std::cout << std::endl;

    return onnx_path;
#else
    throw std::runtime_error("LibTorch not available");
#endif
}

std::string ModelExporter::quantize_onnx_fp16(const std::string& onnx_path) {
    std::string output_path = onnx_path;
    size_t pos = output_path.rfind(".onnx");
    if (pos != std::string::npos) {
        output_path.insert(pos, "_fp16");
    } else {
        output_path += "_fp16.onnx";
    }

    std::cout << "[ModelExporter] FP16 quantization requires onnxconverter-common:" << std::endl;
    std::cout << "  python -c \"from onnxconverter_common import float16; "
              << "import onnx; model = onnx.load('" << onnx_path << "'); "
              << "model_fp16 = float16.convert_float_to_float16(model); "
              << "onnx.save(model_fp16, '" << output_path << "')\"" << std::endl;

    return output_path;
}

void ModelExporter::set_class_names(const std::vector<std::string>& class_names) {
    class_names_ = class_names;
}

void ModelExporter::save_labels(const std::string& model_path) {
    if (class_names_.empty()) return;

    std::string labels_path = fs::path(model_path).replace_extension(".txt").string();
    std::ofstream file(labels_path);
    for (const auto& name : class_names_) {
        file << name << "\n";
    }
    std::cout << "[ModelExporter] Saved labels to " << labels_path << std::endl;
}

// ============================================================================
// Convenience Function
// ============================================================================

std::string export_model(
    const std::string& model_or_checkpoint,
    const std::string& path,
    const ExportOptions& options
) {
    ModelExporter exporter(model_or_checkpoint);
    return exporter.export_model(path, options);
}

} // namespace train
} // namespace ivit
